### PR TITLE
feat: add logits processor support for trtllm backend

### PR DESCRIPTION
**Pull request by bhuvan002**
#### Related Issues: (use one of the action keywords Closes / Fixes / Resolves / Relates to)

- closes GitHub issue: #xxx

## Summary by CodeRabbit

- New Features
  - Support for custom logits processors in the TensorRT-LLM backend, enabling in-place logits modification during generation.
  - Added example processors (Temperature, Hello World) to illustrate usage.

- Chores
  - Tokenizer now initializes by default when starting the TensorRT-LLM engine for improved out-of-the-box readiness.

### CODE REVIEW

File: components/backends/trtllm/src/dynamo/trtllm/logits_processing/adapter.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    def __init__(self, processor: BaseLogitsProcessor):
+        super().__init__()
+        self.processor = processor
+
+    def __call__(
+        self,
+        req_ids: int,
+        logits: torch.Tensor,
+        ids: List[List[int]],
+        stream_ptr,
```
Comment (milesial) [inline components/backends/trtllm/src/dynamo/trtllm/logits_processing/adapter.py:34]: We need to use this stream or else memory dependencies are broken

File: components/backends/trtllm/src/dynamo/trtllm/logits_processing/adapter.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+            for ids_req, logits_req in zip(ids, logits):
+                if logits_req.shape[0] != 1:
+                    raise ValueError(
+                        "Logits processing with beam width > 1 is not supported"
+                    )
+                # Remove dimension 0 from logits_req
+                modified_logits = self.processor(ids_req, logits_req.reshape(-1))
+
+                # TRT-LLM expects in-place modification
+                logits.copy_(modified_logits)
```
Comment (milesial) [inline components/backends/trtllm/src/dynamo/trtllm/logits_processing/adapter.py]: Have you tried putting those two ops in a @torch.compile function to see if we can have a JIT kernel fusing these two? Or some other ways to limit the overhead / number of kernels

Comment (milesial) [inline components/backends/trtllm/src/dynamo/trtllm/logits_processing/adapter.py]: At a minimum, you should move the copy_ outside of the for loop, to have a horizontally fused kernel copying a list of tensors into the batched tensor. See torch foreach_map

File: components/backends/trtllm/src/dynamo/trtllm/main.py [hunk 3]
```diff
@@ -162,7 +162,7 @@ async def init(runtime: DistributedRuntime, config: Config):
         "pipeline_parallel_size": config.pipeline_parallel_size,
         "moe_expert_parallel_size": config.expert_parallel_size,
         "backend": "pytorch",
-        "skip_tokenizer_init": True,
+        "skip_tokenizer_init": False,
```
Comment (milesial) [inline components/backends/trtllm/src/dynamo/trtllm/main.py]: Is this just to unblock examples/tests or does logits processing inherently need the tokenizer on the TRTLLM side?

**Review by milesial — CHANGES_REQUESTED** [review]

File: components/backends/trtllm/src/dynamo/trtllm/logits_processing/adapter.py (hunk 2 above)
Comment (bhuvan002) [reply components/backends/trtllm/src/dynamo/trtllm/logits_processing/adapter.py]: I think it's better if we keep the adapter unopinionated just in case the user provided processor is a pathological case with torch.compile. We can add in our documentation/best practices that the user code itself can add torch.compile if they wish?

File: components/backends/trtllm/src/dynamo/trtllm/main.py (hunk 3 above)
Comment (bhuvan002) [reply components/backends/trtllm/src/dynamo/trtllm/main.py]: No we just need to pass in an interface with a _call_ method. The tokenizer I used is to instantiate the inner HelloWorld processor. For many of the 'interesting' logits processing we do need the tokenizer like for example here https://github.com/NVIDIA/logits-processor-zoo/tree/main/logits_processor_zoo/transformers

File: components/backends/trtllm/src/dynamo/trtllm/logits_processing/adapter.py (hunk 1 above)
Comment (bhuvan002) [reply components/backends/trtllm/src/dynamo/trtllm/logits_processing/adapter.py:34]: Fixed

File: components/backends/trtllm/src/dynamo/trtllm/logits_processing/adapter.py (hunk 2 above)
Comment (bhuvan002) [reply components/backends/trtllm/src/dynamo/trtllm/logits_processing/adapter.py]: Do you think it's better if we instead make the dynamo exposed processors such that the user callback modify the logits in-place and not return anything. And if other backends such as vllm require a returned tensor then we can just return it from the vllm specific adapter instead?

Comment (milesial) [reply components/backends/trtllm/src/dynamo/trtllm/logits_processing/adapter.py]: That might be dangerous if a framework relies on the unprocessed logits and does not expect the in-place modifications. What do vLLM and SGLang do?

Comment (milesial) [reply components/backends/trtllm/src/dynamo/trtllm/logits_processing/adapter.py]: If torch compile is not successful it would fall back to the regular eager execution so it should be fine. I think we should focus on performance by default especially in these core components. Would be bad UX to ask users to modify or extend the code we provide, if they want performance.

File: components/backends/trtllm/src/dynamo/trtllm/main.py (hunk 3 above)
Comment (milesial) [reply components/backends/trtllm/src/dynamo/trtllm/main.py]: Got it, with the tokenizer created at multiple places we must be careful as it would be easy for double-tokenization to happen without us noticing

File: components/backends/trtllm/src/dynamo/trtllm/logits_processing/adapter.py (hunk 2 above)
Comment (bhuvan002) [reply components/backends/trtllm/src/dynamo/trtllm/logits_processing/adapter.py]: We're not supporting sglang at the moment because they don't expose the input token IDs in their API so their logits processors are very limited https://github.com/sgl-project/sglang/issues/3524 
For vllm they support in place (https://github.com/vllm-project/vllm/pull/22919/#discussion_r2292054167) right now they're finalizing the docs but this discussion says in-place is preferred

Comment (bhuvan002) [reply components/backends/trtllm/src/dynamo/trtllm/logits_processing/adapter.py]: Ah I meant the users would have the torch.compile in the user provided custom logits processors, no modifications to the dynamo code. For any logits processors that we officially keep in dynamo like thinking budget early stopping for instance, we can add torch.compile to that separately. But with the fallback behavior I can still wrap this adapter into a torch compile

Comment (bhuvan002) [reply components/backends/trtllm/src/dynamo/trtllm/logits_processing/adapter.py]: Modified it to be in place since both TRTLLM/vllm support in place and cutting down one tensor copy in the adapter

Comment (bhuvan002) [reply components/backends/trtllm/src/dynamo/trtllm/logits_processing/adapter.py]: I tested the hello world processor with torch.compile and it was slower than eager, so I think this decision should be left to the writers of the logits processors.
I also refactored it to expect in place modification thus removing the copy_ kernel from the adapter

Comment (bhuvan002) [reply components/backends/trtllm/src/dynamo/trtllm/logits_processing/adapter.py]: Also since the dynamo BaseLogitsProcessor is per request and we are using the TRTLLM callback also per request in the request sampling params, I removed the batch dimension from the adapter since it was unnecessary. Should be cleaner now.

Comment (milesial) [reply components/backends/trtllm/src/dynamo/trtllm/logits_processing/adapter.py]: Right but since the extra op was in the adapter, then the user-provided callback would not have the opportunity to fuse its ops with the ones the adapter was doing. Now that you removed the copy it should be good but let's keep it in mind if we later add some more logic on the adapter side

Comment (milesial) [reply components/backends/trtllm/src/dynamo/trtllm/logits_processing/adapter.py]: If everyone aligns to inplace then it's ok having the base interface do inplace yes. I don't see any signature change for BaseLogitsProcessor in this MR tho?

Comment (milesial) [reply components/backends/trtllm/src/dynamo/trtllm/logits_processing/adapter.py]: By the way, even for a single request there is a batch dimension in TRTLLM since you can have beam search

Comment (bhuvan002) [reply components/backends/trtllm/src/dynamo/trtllm/logits_processing/adapter.py]: I removed the return type hint from BaseLogitsProcessor here https://github.com/ai-dynamo/dynamo/pull/2702/files#diff-9de0674f878011bb81ace752b6189b39c90163a39b82c523076ff7a5ce5e2b2fR29
Also yes for beam search dimension in TRTLLM, but dynamo E2E doesn't support beam search so I'm just assuming (and enforcing) single sequence here

File: components/backends/trtllm/src/dynamo/trtllm/logits_processing/__init__.py [hunk 4]
```diff
@@ -0,0 +1,6 @@
+# SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+from .adapter import TrtllmDynamoLogitsAdapter, create_trtllm_adapters
+
+__all__ = ["TrtllmDynamoLogitsAdapter", "create_trtllm_adapters"]
```
[the remark this answers is not in the document]
Comment (bhuvan002) [reply components/backends/trtllm/src/dynamo/trtllm/logits_processing/__init__.py]: Not needed. CI fixed by removing __init__.py and moving around imports

File: components/backends/trtllm/src/dynamo/trtllm/logits_processing/adapter.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+import logging
+from typing import List, Optional
+
+import torch
+from tensorrt_llm.sampling_params import LogitsProcessor
+
+from dynamo.logits_processing import BaseLogitsProcessor
+
+logger = logging.getLogger(__name__)
+
```
[the remark this answers is not in the document]
Comment (bhuvan002) [reply components/backends/trtllm/src/dynamo/trtllm/logits_processing/adapter.py:13]: Not needed. CI fixed by removing __init__.py and moving around imports

File: components/backends/trtllm/src/dynamo/trtllm/logits_processing/adapter.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+class TrtllmDynamoLogitsAdapter(LogitsProcessor):
+    """
+    Adapter that wraps Dynamo BaseLogitsProcessor instances to work with TensorRT-LLM's logits processor interface.
+
+    Inherits from tensorrt_llm.LogitsProcessor and implements the required interface:
+    __call__(self, req_ids: int, logits: torch.Tensor, ids: List[List[int]], stream_ptr, client_id: Optional[int])
+
+    This adapter maintains per-request state and converts between the interfaces.
+    """
```
[the remark this answers is not in the document]
Comment (bhuvan002) [reply components/backends/trtllm/src/dynamo/trtllm/logits_processing/adapter.py:23]: Not needed. CI fixed by removing __init__.py and moving around imports

File: components/backends/trtllm/src/dynamo/trtllm/request_handlers/handler_base.py [hunk 7]
```diff
@@ -21,8 +21,10 @@
 from tensorrt_llm import SamplingParams
 from tensorrt_llm.llmapi import DisaggregatedParams as LlmDisaggregatedParams
 
+from dynamo.logits_processing.examples import HelloWorldLogitsProcessor
 from dynamo.runtime.logging import configure_dynamo_logging
 from dynamo.trtllm.engine import TensorRTLLMEngine
+from dynamo.trtllm.logits_processing import create_trtllm_adapters
 from dynamo.trtllm.multimodal_processor import MultimodalRequestProcessor
```
[the remark this answers is not in the document]
Comment (bhuvan002) [reply components/backends/trtllm/src/dynamo/trtllm/request_handlers/handler_base.py]: Not required anymore

File: components/backends/trtllm/src/dynamo/trtllm/request_handlers/handler_base.py [hunk 8]
```diff
@@ -168,6 +170,11 @@
         request_id = request.get("id") or request.get("request_id", "unknown-id")
         model_name = request.get("model", "unknown_model")
 
+        # TODO: Just for testing. Hardcoding the hello world processor.
+        processors = [HelloWorldLogitsProcessor(self.engine.llm.tokenizer)]
+        adapters = create_trtllm_adapters(processors)
+        sampling_params.logits_processor = adapters
+
```
[the remark this answers is not in the document]
Comment (bhuvan002) [reply components/backends/trtllm/src/dynamo/trtllm/request_handlers/handler_base.py]: Added

File: components/backends/trtllm/src/dynamo/trtllm/logits_processing/adapter.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+                        f"This logits adapter only supports per-request logits processing. "
+                        f"Received logits with batch size {logits.shape[0]} expected 1"
+                    )
+                if logits.shape[1] != 1:
+                    raise ValueError(
+                        "Logits processing with beam width > 1 is not supported"
+                    )
+                # Call the processor which modifies the logits in-place
+                self.processor(ids[0], logits[0, 0, :])
+
```
[the remark this answers is not in the document]
Comment (bhuvan002) [reply components/backends/trtllm/src/dynamo/trtllm/logits_processing/adapter.py:64]: Passing None to torch.cuda.stream is a no-op, so this should be fine 
https://docs.pytorch.org/docs/stable/generated/torch.cuda.stream.html#torch.cuda.stream

File: components/backends/trtllm/src/dynamo/trtllm/main.py (hunk 3 above)
Comment (tanmayv25) [inline components/backends/trtllm/src/dynamo/trtllm/main.py]: Why setting this to False? Is it only needed when custom logits processing is provided? 
See L213 where we set it to False for multimodal case. Should this be conditionalized as well instead of making it `False` by default. 

We set `default_sampling_params.detokenize` to False in L226, so we may be avoiding double detokenization. However, did you verify the same?

**Review by tanmayv25 — COMMENTED** [review]
Add some documentation about this feature in backend readme?

**Comment by bhuvan002** [comment]
> Add some documentation about this feature in backend readme?

@tanmayv25 Added docs to the trtllm readme

File: components/backends/trtllm/src/dynamo/trtllm/main.py (hunk 3 above)
Comment (bhuvan002) [reply components/backends/trtllm/src/dynamo/trtllm/main.py]: Made it conditional with the feature/debug flag. And no double detokenizing is happening, the rust engine still gets the returned token ids and does the detokenizing. The inited tokenizer is only passed to the logits processor to get model specific tokens.
I can also change it so instead of modifying `skip_tokenizer_init` I instead use the tokenizer object in L222 using `tokenizer = tokenizer_factory(arg_map["model"])` and then pass that as a request handler config param so we get that in the BaseHandler instead of accessing it through `self.engine.llm.tokenizer` if that is better?

**Review by tanmayv25 — APPROVED** [review]
